File: client/python/examples/cancelling.py
"""
A full example of cancelling jobs, either with their job id, or
the job-set id.
"""

import os
import uuid

import grpc
from armada_client.client import ArmadaClient
from armada_client.k8s.io.api.core.v1 import generated_pb2 as core_v1
from armada_client.k8s.io.apimachinery.pkg.api.resource import (
    generated_pb2 as api_resource,
)
from armada_client.typings import JobState


def create_dummy_job(client: ArmadaClient):
    """
    Create a dummy job with a single container.
    """

    # For infomation on where this comes from,
    # see https://github.com/kubernetes/api/blob/master/core/v1/generated.proto
    pod = core_v1.PodSpec(
        containers=[
            core_v1.Container(
                name="container1",
                image="index.docker.io/library/ubuntu:latest",
                args=["sleep", "10s"],
                securityContext=core_v1.SecurityContext(runAsUser=1000),
                resources=core_v1.ResourceRequirements(
                    requests={
                        "cpu": api_resource.Quantity(string="120m"),
                        "memory": api_resource.Quantity(string="510Mi"),
                    },
                    limits={
                        "cpu": api_resource.Quantity(string="120m"),
                        "memory": api_resource.Quantity(string="510Mi"),
                    },
                ),
            )
        ],
    )

    return [client.create_job_request_item(priority=1, pod_spec=pod)]


def cancelling_jobs_example(client, queue):
    """
    Submits and Cancels Jobs by two different methods:

    - Cancelling a job by its job id
    - Cancelling a job by its job set id
    """

    # Create the PodSpec for the job
    job_request_items1 = create_dummy_job(client)
    job_request_items2 = create_dummy_job(client)

    job_set_id1 = f"set-{uuid.uuid1()}"
    job_set_id2 = f"set-{uuid.uuid1()}"

    resp1 = client.submit_jobs(
        queue=queue, job_set_id=job_set_id1, job_request_items=job_request_items1
    )

    # Gets the job_id of the first job in job_request_items1
    # This job is cancelled using the job_id
    job_id = resp1.job_response_items[0].job_id
    client.cancel_jobs(job_id=job_id)

    client.submit_jobs(
        queue=queue, job_set_id=job_set_id2, job_request_items=job_request_items2
    )

    # This job is cancelled using the queue and job_set
    client.cancel_jobs(queue=queue, job_set_id=job_set_id1)


def cancelling_jobset_with_filter(client, queue):
    """
    Cancels a jobset with a filter.
    """

    # Create the PodSpec for the job
    job_request_items = create_dummy_job(client)

    # Create the jobset with a filter
    job_set_id = f"set-{uuid.uuid1()}"
    client.submit_jobs(
        queue=queue, job_set_id=job_set_id, job_request_items=job_request_items
    )

    # Cancel the jobset with a filter
    client.cancel_jobset(
        queue=queue,
        job_set_id=job_set_id,
        filter_states=[JobState.PENDING, JobState.RUNNING],
    )


def quick_create_queue(client, queue):
    """
    Creates a queue.

    Will skip if the queue already exists.
    """

    queue_req = client.create_queue_request(name=queue, priority_factor=1)

    # Make sure we handle the queue already existing
    try:
        client.create_queue(queue_req)

    # Handle the error we expect to maybe occur
    except grpc.RpcError as e:
        code = e.code()
        if code == grpc.StatusCode.ALREADY_EXISTS:
            print(f"Queue {queue} already exists")
            client.update_queue(queue_req)
        else:
            raise e


def workflow():
    """
    Starts a workflow, which includes:
        - Creating a queue and job
        - Cancelling a job with its job-id and jobset-id
        - Cancelling a job-set with a filter
    """

    # The queue and job_set_id that will be used for all jobs
    queue = "test-cancelling"

    # Ensures that the correct channel type is generated
    if DISABLE_SSL:
        channel = grpc.insecure_channel(f"{HOST}:{PORT}")
    else:
        channel_credentials = grpc.ssl_channel_credentials()
        channel = grpc.secure_channel(
            f"{HOST}:{PORT}",
            channel_credentials,
        )

    client = ArmadaClient(channel)
    quick_create_queue(client, queue)

    cancelling_jobs_example(client, queue)
    cancelling_jobset_with_filter(client, queue)


if __name__ == "__main__":
    # Note that the form of ARMADA_SERVER should be something like
    # domain.com, localhost, or 0.0.0.0
    DISABLE_SSL = os.environ.get("DISABLE_SSL", False)
    HOST = os.environ.get("ARMADA_SERVER", "localhost")
    PORT = os.environ.get("ARMADA_PORT", "50051")

    workflow()
    print("Completed Workflow")

File: client/rust/examples/cancelling.rs
use anyhow::Result;
use armada_client::{
    armada::{
        api::{JobState, JobSubmitRequestItem, Queue},
        k8s::io::{
            api::core::v1::{Container, PodSpec, ResourceRequirements, SecurityContext},
            apimachinery::pkg::api::resource::Quantity,
        },
    },
    ArmadaClient,
};
use tonic::transport::Channel;

fn create_dummy_job() -> Vec<JobSubmitRequestItem> {
    let mut pod_spec = PodSpec::default();
    let mut container = Container::default();
    container.name = Some(String::from("container1"));
    container.image = Some("index.docker.io/library/ubuntu:latest".to_string());
    container.args = vec![String::from("sleep"), String::from("10s")];
    container.security_context = Some({
        let mut sec_ctx = SecurityContext::default();
        sec_ctx.run_as_user = Some(1000);
        sec_ctx
    });
    container.resources = Some({
        let mut reqs = ResourceRequirements::default();
        reqs.requests.insert(
            String::from("cpu"),
            Quantity {
                string: Some(String::from("120m")),
            },
        );
        reqs.requests.insert(
            String::from("memory"),
            Quantity {
                string: Some(String::from("510Mi")),
            },
        );

        reqs.limits.insert(
            String::from("cpu"),
            Quantity {
                string: Some(String::from("120m")),
            },
        );
        reqs.limits.insert(
            String::from("memory"),
            Quantity {
                string: Some(String::from("510Mi")),
            },
        );
        reqs
    });
    pod_spec.containers = vec![container];

    let mut jsr = JobSubmitRequestItem::default();
    jsr.priority = 1.0;
    jsr.pod_specs = vec![pod_spec];
    jsr.namespace = "personal-anonymous".to_string();
    vec![jsr]
}

async fn create_queue(client: &mut ArmadaClient, queue: &str) -> Result<()> {
    let mut queue_req = Queue::default();
    queue_req.name = queue.to_string();
    queue_req.priority_factor = 1.0;
    client.create_queue(queue_req).await?;
    println!("Created queue: {}", queue);
    Ok(())
}

async fn cancel_jobs(client: &mut ArmadaClient, queue: &str) -> Result<()> {
    let request_items1 = create_dummy_job();
    let request_items2 = create_dummy_job();

    let job_set_id1 = format!("simple-jobset-{}", uuid::Uuid::new_v4());
    let job_set_id2 = format!("simple-jobset-{}", uuid::Uuid::new_v4());

    let resp1 = client
        .submit_jobs(queue, &job_set_id1, request_items1)
        .await?;

    println!("Submitted job set: {}", job_set_id1);

    let job_id = &resp1
        .get_ref()
        .job_response_items
        .get(0)
        .expect("Failed to get the job response")
        .job_id;

    client.cancel_jobs("", job_id, "").await?;

    client
        .submit_jobs(queue, &job_set_id2, request_items2)
        .await?;

    println!("Submitted job set: {}", job_set_id2);

    client.cancel_jobs(queue, "", &job_set_id2).await?;

    Ok(())
}

async fn cancel_jobset_with_filter(client: &mut ArmadaClient, queue: &str) -> Result<()> {
    let request_items1 = create_dummy_job();

    let job_set_id1 = format!("simple-jobset-{}", uuid::Uuid::new_v4());

    client
        .submit_jobs(queue, &job_set_id1, request_items1)
        .await?;

    println!("Submitted job set: {}", job_set_id1);

    client
        .cancel_jobset(
            queue,
            &job_set_id1,
            vec![JobState::Pending, JobState::Running],
        )
        .await?;
    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    let disable_ssl = std::env::var("DISABLE_SSL").map_or(false, |_| true);
    let host = std::env::var("ARMADA_SERVER").unwrap_or(String::from("localhost"));
    let port: u32 = std::env::var("ARMADA_PORT").map_or(50051, |p| p.parse::<u32>().unwrap());

    let channel = if disable_ssl {
        Channel::from_shared(format!("http://{host}:{port}"))?
            .connect()
            .await?
    } else {
        Channel::from_shared(format!("https://{host}:{port}"))?
            .connect()
            .await?
    };
    let mut client = ArmadaClient::new(channel);
    let queue = format!("simple-queue-{}", uuid::Uuid::new_v4());

    create_queue(&mut client, &queue).await?;
    cancel_jobs(&mut client, &queue).await?;
    cancel_jobset_with_filter(&mut client, &queue).await?;

    Ok(())
}
